Modules/TPC/src/TrendingTaskTPC.cxx

```diff
@@ -0,0 +1,233 @@
+{
+  mConfig = TrendingTaskConfigTPC(name, config);
+}
+
+void TrendingTaskTPC::initialize(Trigger, framework::ServiceRegistry&)
+{
+  // Prepare the data structure of TTree.
+  mTrend = std::make_unique<TTree>(); // TODO: retrieve last TTree, so we continue trending. maybe do it optionally?
+  mTrend->SetName(PostProcessingInterface::getName().c_str());
+  mTrend->Branch("meta", &mMetaData, "runNumber/I");
+  mTrend->Branch("time", &mTime);
+
+  // Define the branch for each datasource (TH1, TH2, canvas,...) in the config.
+  for (const auto& source : mConfig.dataSources) {
+    //gROOT->ProcessLine("#include <vector>"); // Check if this is needed when reductor has vector branch
+    std::unique_ptr<ReductorTPC> reductor(root_class_factory::create<ReductorTPC>(source.moduleName, source.reductorName));
+    mTrend->Branch(source.name.c_str(), reductor->getBranchAddress(), reductor->getBranchLeafList());
+    mReductors[source.name] = std::move(reductor);
+  }
+
+  // Initialise the publishing by getting the current mTrend TTree.
+  getObjectsManager()->startPublishing(mTrend.get());
+}
+
+// Ttodo: see if OptimizeBaskets() indeed helps after some time.
+void TrendingTaskTPC::update(Trigger t, framework::ServiceRegistry& services)
+{
+  auto& qcdb = services.get<repository::DatabaseInterface>();
+  trendValues(t.timestamp, qcdb);
+  generatePlots();
+}
+
+void TrendingTaskTPC::finalize(Trigger, framework::ServiceRegistry&)
+{
+  generatePlots();
+}
+
+void TrendingTaskTPC::trendValues(uint64_t timestamp, repository::DatabaseInterface& qcdb)
+{
+  mTime = timestamp / 1000; // ROOT expects seconds since epoch.
+  // Todo get run number when it is available. consider putting it inside monitor object's metadata (this might be not enough if we trend across runs).
+  mMetaData.runNumber = -1;
+
+  for (auto& dataSource : mConfig.dataSources) {
+    // Todo: make it agnostic to MOs, QOs or other objects. Let the reductor cast to whatever it needs.
+
+    // Check if the dataSource is a canvas (name starts with "c_") or not.
+    // Return true if it is a canvas.
+    bool isDataCanvas = boost::algorithm::starts_with(dataSource.name, "c_");
+
+    // Get the size of the axisDivision for the number of pads to prepare.
+    int axisSize = (int)dataSource.axisDivision.size();
+    int innerAxisSize = (int)dataSource.axisDivision[0].size() - 1;
+    mNumberPads = axisSize * innerAxisSize;
+
+    if (dataSource.type == "repository") {
+      auto mo = qcdb.retrieveMO(dataSource.path, dataSource.name, timestamp);
+      TObject* obj = mo ? mo->getObject() : nullptr;
+      if (obj) {
+        mReductors[dataSource.name]->update(obj, dataSource.axisDivision, isDataCanvas);
+
+        // Get the number of input pads if obj is a TCanvas. Update then the value obtained above.
+        if (isDataCanvas) {
+          auto canvas = dynamic_cast<TCanvas*>(obj);
+          mNumberPads = dynamic_cast<TList*>(canvas->GetListOfPrimitives())->GetEntries();
+        }
```

Comment:
```suggestion
        // Get the number of input pads if obj is a TCanvas. Update then the value obtained above.
        if (auto canvas = dynamic_cast<TCanvas*>(obj);) {
          mNumberPads = static_cast<TList*>(canvas->GetListOfPrimitives())->GetEntries();
        }
```

Reply:
Here the `isDataCanvas` is anyhow only used once, right? Then I would follow this suggestion here.